File: Node/Euler014.js
#!/usr/bin/env node

function main() {
  const target = 1000000;
  var biggestChain = -1;
  var biggestStart = -1;

  for (var i = 1; i < target; i++) {
    l = collatz(i).length;
    if (l > biggestChain) {
      biggestChain = l;
      biggestStart = i;
    }
  }

  console.log(biggestStart);
}

function collatz(n) {
  var result = [];

  while (n > 1) {
    result.push(n);
    if (n % 2 == 0) {
      n /= 2;
    } else {
      n = 3 * n + 1;
    }
  }

  result.push(1);
  return result;
}

if (require.main == module) {
  main();
}

File: Ruby/Euler014.rb
#!/usr/bin/env ruby

def main
  target = 1000000

  biggest_chain = -1
  biggest_num = -1
  for i in (1..target)
    chain = collatz(i)
    if chain.length > biggest_chain
      biggest_chain = chain.length
      biggest_num = i
    end
  end

  puts biggest_num
end

def collatz(n)
  result = []

  while n > 1
    result.append(n)
    if n % 2 == 0
      n /= 2
    else
      n = 3 * n + 1
    end
  end

  result.append(1)
  return result
end

if __FILE__ == $0
  main
end
